Reduce number of Window test configurations

malfet opened the pull request:
After this diff, on PR following compilation configuration would be running:
     - VS2017 14.11, CUDA10.1
     - VS2017 no CUDA, CUDA10.1
     - VS2019, CUDA10.1
    And tested:
     - VS2017 14.11, CUDA10.1
     - VS2017 14.11 no CUDA (only 1st half of tests)
     - VS2017 14.11 force on CPU (only 1st half of test)
    
    And on master, we would be building both VS2017 14.11 and 14.16, but testing only VS2017 14.11 and VS2019 builds.

malfet committed:
Reduce number of Window test configurations

After this diff, on PR following compilation configuration would be running:
 - VS2017 14.11, CUDA10.1
 - VS2019, CUDA10.1
 - VS2019, no CUDA
And tested:
 - VS2019 CUDA10.1
 - VS2019 no CUDA (only 1st half of tests)
 - VS2019 CUDA force on CPU (only 1st half of test)

And on master, we would be building both VS2017 14.11 and 14.16, but testing only VS2019 builds and smoke testing VS2017 14.11

peterjc123 commented on .circleci/config.yml:
```diff
@@ -2083,43 +2077,20 @@ workflows:
           vc_version: "14.11"
           vc_year: "2017"
-      - pytorch_windows_test:
-          build_environment: pytorch-win-vs2017-14-11-cuda10-cudnn7-py3
```
Wow, removing the test jobs here makes the library totally untested on 14.11/14.16.

ezyang replied:
I think @malfet's contention here is that most of the signal we're getting from the Windows builds is flushing out MSVC compiler bugs. Do you think there's also runtime bugs that are likely to crop up on 14.11/14.16 but not on the most recent one?

malfet replied:
If I understand correctly, the binary that we are published is build using VS2019, and it seems to work on all currently supported Windows platforms, isn't it?
In that case, what is the point of testing binaries build with older versions of VisualStudio, if they are later not being used?

peterjc123 replied:
@malfet No, the binaries are built with MSVC 2017 14.16 (except CUDA 9.2 binaries which use 14.11).

peterjc123 replied:
I hope that we can migrate to VS 2019 but this means that we need to deprecate support for CUDA <=10.0.

malfet replied:
@peterjc123 is that due to the Python binary runtime requirements? Or just historically?

peterjc123 replied:
> I think @malfet's contention here is that most of the signal we're getting from the Windows builds is flushing out MSVC compiler bugs. Do you think there's also runtime bugs that are likely to crop up on 14.11/14.16 but not on the most recent one?

No, we don't need to ensure PyTorch builds on every version of VC toolchain. But at least, it should work on the toolchains that we used in binary builds.

malfet replied:
> No, we don't need to ensure PyTorch builds on every version of VC toolchain. But at least, it should work on the toolchains that we used in binary builds.
If VS2019 results in faster/smaller binaries, why not use it for binary builds?

I also wonder if CUDA-9.2 builds on Windows can be dropped in time for next release.

peterjc123 replied:
@malfet Copied from README.md
CUDA version | Newest supported VS version
-- | --
9.0 / 9.1 | Visual Studio 2017 Update 4 (15.4) (_MSC_VER <= 1911)
9.2 | Visual Studio 2017 Update 5 (15.5) (_MSC_VER <= 1912)
10.0 | Visual Studio 2017 (15.X) (_MSC_VER < 1920)
10.1 | Visual Studio 2019 (16.X) (_MSC_VER < 1930)

Old CUDA relies on old VS.

peterjc123 replied:
> > No, we don't need to ensure PyTorch builds on every version of VC toolchain. But at least, it should work on the toolchains that we used in binary builds.
> > If VS2019 results in faster/smaller binaries, why not use it for binary builds?
> 
> I also wonder if CUDA-9.2 builds on Windows can be dropped in time for next release.

Well, I guess we can drop the support once CUDA 11 is out in Anaconda Cloud.

peterjc123 replied:
There is already a RFC for deprecating CUDA 9.2. https://github.com/pytorch/pytorch/issues/26646 I'm okay with removing the 14.11 build/test jobs once we remove CUDA 9.2 support.

malfet replied:
Looks like two topics are being discussed here:
 - CUDA support matrix
 - What is currently tested in OSS CI
Leaving first question aside, I still think we don't get much extra signal from testing PyTorch build with  CUDA-10.1 and 3 different versions of Visual Studio. And this is what I'm proposing to  do with this PR: change build/test matrix to cover different compiler builds but test only one out of 3 configurations, since, to the best of my knowledge, there are very few examples of PyTorch runtime behavior is drastically different(i.e. test fails in one config but works in another) between binaries compiled using different version of compiler.

peterjc123 replied:
@malfet Yes, maybe you are right. But actually the build doesn't cover all the codepaths. I strongly suggest we should still test building the c++ extensions at least.

malfet replied:
Update `binary_windows_builds` to use VS2019

peterjc123 commented on .circleci/scripts/binary_windows_build.sh:
```diff
@@ -5,7 +5,7 @@ source "/c/w/env"
 
 export CUDA_VERSION="${DESIRED_CUDA/cu/}"
-export VC_YEAR=2017
+export VC_YEAR=2019
```
Wait, before merging this PR, some changes are needed at the builder side. Otherwise, the conda packages will be broken.

peterjc123 replied:
I think it's better to do that in a separate PR. Let me do this for you.

peterjc123 replied:
@malfet The required changes for builder is in https://github.com/pytorch/builder/pull/445. And here I guess you'll need to make a bit modification.
```suggestion
if [[ "$CUDA_VERSION" == "92" || "$CUDA_VERSION" == "100" ]]; then
  export VC_YEAR=2017
else
  export VC_YEAR=2019
fi
```

malfet replied:
Actually, we can't make this change, as by convention all Python extensions on Windows must be built with VS2017:
https://docs.conda.io/projects/conda-build/en/latest/user-guide/tutorials/building-conda-packages.html

peterjc123 replied:
@malfet Well, you can. But you'll need to write a custom `conda_build_config.yaml`. I've already done this for torchaudio and torchtext.

malfet replied:
Do VS2017 and VS2019 have binary compatible C++ ABI? If not it might cause subtle runtime failures, even if builds and smoke testing succeeds.

peterjc123 replied:
See https://devblogs.microsoft.com/cppblog/cpp-binary-compatibility-and-pain-free-upgrades-to-visual-studio-2019/. It's safe to link an executable with libraries generated by VC 14.11 / 14.16  and 14.24. But it won't work if you use /GL (whole program optimization).

malfet committed:
Add `test_cpp_extensions` to test1 part of subtests

CHange test_python_all_except_nn to test_python_all_except_nn_and_cpp_extensions
Call test_cpp_extensions explicitly
